src/goto-instrument/code_contracts.cpp

```diff
@@ -961,3 +1005,713 @@ bool code_contractst::enforce_contracts(
+
+goto_programt &assigns_clauset::temporary_declarations(
+  source_locationt location,
+  irep_idt function_name,
+  irep_idt language_mode)
+{
+  if(standin_declarations.empty())
+  {
+    for(assigns_clause_targett *target : targets)
+    {
+      for(symbol_exprt symbol : target->temporary_declarations())
+      {
+        standin_declarations.add(
+          goto_programt::make_decl(symbol, symbol.source_location()));
+      }
+    }
+  }
+  return standin_declarations;
+}
+
+goto_programt assigns_clauset::dead_stmts(
+  source_locationt location,
+  irep_idt function_name,
+  irep_idt language_mode)
+{
+  goto_programt dead_statements;
+  for(assigns_clause_targett *target : targets)
+  {
+    for(symbol_exprt symbol : target->temporary_declarations())
+    {
+      dead_statements.add(
+        goto_programt::make_dead(symbol, symbol.source_location()));
+    }
+  }
+  return dead_statements;
+}
+
+goto_programt assigns_clauset::havoc_code(
+  source_locationt location,
+  irep_idt function_name,
+  irep_idt language_mode)
+{
+  goto_programt havoc_statements;
+  for(assigns_clause_targett *target : targets)
+  {
+    for(goto_programt::instructiont instruction :
+        target->havoc_code(location).instructions)
+    {
+      havoc_statements.add(std::move(instruction));
+    }
+  }
+  return havoc_statements;
+}
+
+exprt assigns_clauset::alias_expression(const exprt &lhs)
+{
+  if(targets.empty())
+  {
+    return false_exprt();
+  }
+
+  exprt left_ptr = assigns_clause_targett::pointer_for(lhs);
+
+  bool first_iter = true;
+  exprt result = false_exprt();
+  for(assigns_clause_targett *target : targets)
+  {
+    if(first_iter)
+    {
+      result = target->alias_expression(left_ptr);
+      first_iter = false;
+    }
+    else
+    {
+      result = or_exprt(result, target->alias_expression(left_ptr));
+    }
+  }
+  return result;
+}
+
+exprt assigns_clauset::compatible_expression(
+  const assigns_clauset &called_assigns)
+{
+  if(called_assigns.targets.empty())
+  {
+    return true_exprt();
+  }
+
+  bool first_clause = true;
+  exprt result = true_exprt();
+  for(assigns_clause_targett *called_target : called_assigns.targets)
+  {
+    bool first_iter = true;
+    exprt current_target_compatible = false_exprt();
+    for(assigns_clause_targett *target : targets)
+    {
+      if(first_iter)
+      {
+        current_target_compatible =
+          target->compatible_expression(*called_target);
+        first_iter = false;
+      }
+      else
+      {
+        current_target_compatible = or_exprt(
+          current_target_compatible,
+          target->compatible_expression(*called_target));
+      }
+    }
+    if(first_clause)
+    {
+      result = current_target_compatible;
+      first_clause = false;
+    }
+    else
+    {
+      exprt::operandst conjuncts;
+      conjuncts.push_back(result);
+      conjuncts.push_back(current_target_compatible);
+      result = conjunction(conjuncts);
+    }
+  }
+
+  return result;
+}
```

Comment:
A. Please can these changes be broken up a bit.
B. Adding 70% more code to this file is quite a bit.  Is there an obvious way of splitting this up / out?